graph/validation_barrier.go

```diff
@@ -210,58 +285,176 @@ func (v *ValidationBarrier) WaitForDependants(job interface{}) error {
+
+		case <-jobChan:
+			// Every time this is sent on or if it's closed, a
+			// parent job has finished. The parent jobs have to
+			// also potentially close the channel because if all
+			// the parent jobs finish and call SignalDependents
+			// before the goroutine running WaitForParents has a
+			// chance to grab the notification chan from
+			// childJobChans, then the running goroutine will wait
+			// here for a notification forever. By having the last
+			// parent job close the notificiation chan, we avoid
+			// this issue.
+
+			// Check and see if we have any parent jobs left. If we
+			// don't, we can finish up.
+			v.Lock()
+			info, found := v.jobInfoMap[annID]
+			if !found {
+				v.Unlock()
+
+				// No parent job info found, proceed with
+				// validation.
+				return nil
+			}
+
+			x := parentJobIDs.Intersect(info.activeParentJobIDs)
+			v.Unlock()
+			if x.Empty() {
+				// The parent jobs have all completed. We can
+				// proceed with validation.
+				return nil
+			}
+
+			// If we've reached this point, we are still waiting on
+			// a parent job to complete.
+		}
+	}
 }
 
-// SignalDependants will allow/deny any jobs that are dependent on this job that
-// they can continue execution. If the job doesn't have any dependants, then
-// this function sill exit immediately.
-func (v *ValidationBarrier) SignalDependants(job interface{}, allow bool) {
+// SignalDependents signals to any child jobs that this parent job has
+// finished.
+func (v *ValidationBarrier) SignalDependents(job interface{}, id JobID) error {
 	v.Lock()
 	defer v.Unlock()
 
 	switch msg := job.(type) {
-
-	// If we've just finished executing a ChannelAnnouncement, then we'll
-	// close out the signal, and remove the signal from the map of active
-	// ones. This will allow/deny any dependent jobs to continue execution.
 	case *lnwire.ChannelAnnouncement1:
-		finSignals, ok := v.chanAnnFinSignal[msg.ShortChannelID]
-		if ok {
-			if allow {
-				close(finSignals.allow)
-			} else {
-				close(finSignals.deny)
-			}
-			delete(v.chanAnnFinSignal, msg.ShortChannelID)
+		// Signal to the child jobs that parent validation has
+		// finished. We have to call removeParentJob for each annID
+		// that this ChannelAnnouncement can be associated with.
+		err := v.removeParentJob(msg.ShortChannelID, id)
+		if err != nil {
+			return err
+		}
+
+		err = v.removeParentJob(route.Vertex(msg.NodeID1), id)
+		if err != nil {
+			return err
+		}
+
+		err = v.removeParentJob(route.Vertex(msg.NodeID2), id)
+		if err != nil {
+			return err
 		}
 
-		delete(v.chanEdgeDependencies, msg.ShortChannelID)
+		return nil
 
-	// For all other job types, we'll delete the tracking entries from the
-	// map, as if we reach this point, then all dependants have already
-	// finished executing and we can proceed.
 	case *lnwire.NodeAnnouncement:
-		delete(v.nodeAnnDependencies, route.Vertex(msg.NodeID))
+		// Remove child job info.
+		v.removeChildJob(route.Vertex(msg.NodeID), id)
```

Comment:
I don't think we need two distinct removal functions for parent and child.

Reply:
changed, using a bool now since they have different reuqirements